### ISSUE TITLE
[Bug]: Force loaded chunks not unloading, even when offline chunkloading=false

### ISSUE THREAD

**Issue by Raidobw2** [FTBTeam/FTB-Mods-Issues#101]
### Mod

FTB Chunks

### Mod version

ftb-chunks-forge-1605.3.2-build.71.jar

### Forge / Fabric version

1.16.5 36.2.8

### Modpack & version

FTB OceanBlock 1.5.0 & FTB University 1.16 1.3.0

### What issue are you having?

Chunks marked as force loaded by players will not unload. This has been a consistent issue since the packs were released and I've had to unload players' chunks manually every so often. I think the issue is they fail to unload for some reason.

This is very bad as lag will pile up on a server as more and more people join and forceload chunks.

### Steps to reproduce

1. Go on a server that has players with force loaded chunks.
2. Use Crash Utilities mod
3. Press CTRL+U to bring up the menu.
4. Click the second tab for a list.
5. Click the "ENTITIES" button and switch to "TICKETS".
6. See list of forced tickets by clicking on the text.

### Anything else to note?

Expected behavior: Chunks should unload
Linked by main source, OceanBlock: FTBTeam/FTB-Modpack-Issues#231
Linked to secondary source, University: FTBTeam/FTB-Modpack-Issues#236

**Comment by fabrossmann** [comment]
I can also confirm this issue on ATM 6. 
https://i.imgur.com/oCZEupy.png

Using the mod observable also shows the loaded chunks in the profiler

**Comment by legitswitch** [comment]
Still a confirmed issue up to build 79, not tested past that.

**Comment by fabrossmann** [comment]
> Are you using Dynamic View mod by any chance?
I am. But this shouldn't be a issue regarding the perma loading. We also had the Issue on FTB Endeavour, where this mod is not included

**Comment by ButrBwedd** [comment]
I'm having the exact opposite issue. There seems to be a bug that when the player disconnects, it unloads the chunks (this is intentional and set in the config files) but when the player rejoins, the chunks arent loaded again. The menu says they are loaded but updates when you toggle force loading on a chunk, whether it used to be loaded prior or not. This is an issue for the fact that you have to manually toggle force loading in every chunk you had loaded every time you join. I cannot seem to find what is making this happen.

**Comment by DeadPsych** [comment]
Unfortunately, I feel the same way, I can only agree with the previous speakers.

Maybe we could find out whether the problem is being worked on there?

Thank you for the effort!

**Comment by Seniorendi** [comment]
Updates?

**Comment by Duckgrass** [comment]
With the jar installed, crashed utils shows that almost every base has its chunks loaded by a forge:entity_ticking. Removing the mod jar causes this issue to no longer happen.
![image](https://user-images.githubusercontent.com/68453016/171565004-4792373e-55d2-48cc-96c9-0a235488e8b1.png)

**Comment by Diagnul** [comment]
I also ran into this issue in the FTB Plexiglass Mountain pack. Some force loaded chunks never unload even when told to unload in the GUI. Unloading the chunks with commands did not work and removing the `minecraft/world/ftbchunks` directory did not fix it either. The mod reports that no chunks are loaded but Crash Utilities still shows all the chunks as loaded and ticking. Removing the FTB Chunks mod resolved the issue but I wished to keep the chunk claiming feature so this was undesirable.

I was able to work around the issue by removing the `chunks.dat` file located in `minecraft/world/data/`. That file appears to store the data for which chunks should be force loaded. Removing it before starting the server prevented all of the forced chunks from being loaded without affecting claimed chunks.

**Comment by jinkhya** [comment]
Hello, 
I also noticed that in various 1.16 and 1.18 packs. I tried forcing it with ftb ranks with the permission node to false as well but no luck. I tried using limited chunkloading mod which does unload chunks but upon player relog the chunks don't re-chunkload so you have to unmark/mark chunks as chunkloading every time, which isn't ideal.

This is really problematic for servers with a lot of different players playing at different times. I hope this issue gets some love soon-ish (as much as life permits ofc) so we can all have a better experience on servers <3
 
Thank you!
Jin

**Comment by sirus628** [comment]
> With the jar installed, crashed utils shows that almost every base has its chunks loaded by a forge:entity_ticking. Removing the mod jar causes this issue to no longer happen. ![image](https://user-images.githubusercontent.com/68453016/171565004-4792373e-55d2-48cc-96c9-0a235488e8b1.png)

ATM7 4.23 Forge 40.1.60 has this issue as well. deleting chunks.dat was working to a degree before 4.23 but it seems like doing that makes the server unable to boot with out crashing now (hangs indefinitely right after loading spawn region).  

Around 400-1000 forge:entity tickets up on chunks at any given time is even giving a ryzen9 3900x with access to all cores a hard time. Right now I am looking for any way to unload these chunks with out removing the claims.

**Comment by MichaelHillcox** [comment]
Please note, I am aware of this issue and I'm trying to find some time to sort it out

**Comment by togreglove** [comment]
Some things I've noticed. Need to do a more clean test.
If you claim things as an individual player and chunkload them, then join an ftb team, you can't access the chunks anymore and have to leave the team, unload the chunks, and rejoin the team.
However if you are on a team, claim and chunkload some chunks, then quit, it shows the chunks as unclaimed, but they are still ticking, even offline, and even if you reform the team, you can't stop them. You can reclaim them and chunkload/unchunkload, but they will tick forever.
I have found no in-game commands that would affect the chunkloaded state of blocks in that condition. Unload says it's not claimed.
Over time, this causes small areas of the map to become perma-chunkloaded and causes issues.

**Comment by jeremiahwinsley** [comment]
We noticed on our servers that clearing the Forge chunks.dat seems to resolve the problem, at least temporarily. It looks like FTB Chunks is using the Forge chunk manager, but not providing the callback to remove expired chunk loads (since Forge and FTB Chunks both persist loaded chunks). It could likely just use the vanilla chunk manager here like it does on Fabric, since that doesn't persist across restart. This would also allow using a custom ticket instead of forge:entity.

**Comment by phit** [comment]
this is such a meme at this point since it hasn't worked for years, that there's other mods adding workarounds for this

https://github.com/Darkere/CrashUtilities/blob/7d57f501bd3b96853a37af8783c57246991866f4/src/main/java/com/darkere/crashutils/CrashUtils.java#L150-L168

as far as I remember this didn't even work properly most of the time when this was still called FTB Utilities

**Comment by jinkhya** [comment]
Yes but I believe players have to manually mark all their chunks as chunkloading every time they log back on with that option of crashutilities on. (which can be tedious if you have to go to 6 different places, sometimes in other dimensions too, to do it)

**Comment by Raidobw2** [comment]
I've poked my mod development team about it, thanks all for commenting. I'll try and have the issue updated soon / add more info here

**Comment by jeremiahwinsley** [comment]
I don't have a consistent method of reproducing this at the moment, I'll take a look later, but as I mentioned above we saw this happen when the FTB Chunks loads get desynced with the Forge loads (because it saves them in two places). This happens frequently on our public servers, possibly to do with joining/leaving teams and force loads not being removed completely.

**Comment by desht** [comment]
OK, thanks.  I might be able to do something useful with Forge's validation callback (i.e. ensure any forge-loaded chunks are also still listed as loaded by FTB Chunks).

Note this won't be relevant on 1.16.5 (where others have reported problems), since the 1.16.5 version of the mod doesn't use the Forge chunkloader functionality *at all*.  It's purely vanilla forceloading there.

**Comment by togreglove** [comment]
The way I cheeze the bug to keep my chunks loaded is to join the server, claim and force-load chunks, create a team, realize I lost access to my old chunks because I'm now in a team, quit the team, unload the chunks, create the team again.  Most of the time, those chunks that were force-loaded when I formed the team are now perma-loaded. (I may have gotten it backwards, it might be that chunks you forceload while in a team get bugged when you leave the team and leave the team empty. I am not online to test this right now)

**Comment by jinkhya** [comment]
I've noticed that upon server restart, a lot of chunks are still loaded even when nobody else has joined yet. As an example, on ATM7 I played a little, made some bonsai trees outputing in really big sophisticated chests. Stopped playing for a couple weeks, and the other day when I joined back, all the chests were completely full with like more than 10 thousands stacks in them. Nobody in my team and no neighbors around to load it either.
I'm not sure when players are kicked with the idle setting in server.properties that it unloads either.

**Comment by desht** [comment]
Thanks for the info, all.  Just to let you know, I'm working through this now - spotted a few issues already (mainly around the creation and deletion of parties, and the transfer of claims between players and parties). Hopefully will have something soon(ish) but there's a lot of corner cases to test here - I want to get this as robust as possible.

**Comment by Malaxis12** [comment]
> Thanks for the info, all. Just to let you know, I'm working through this now - spotted a few issues already (mainly around the creation and deletion of parties, and the transfer of claims between players and parties). Hopefully will have something soon(ish) but there's a lot of corner cases to test here - I want to get this as robust as possible.

hello, how to update files to fix issue?

**Comment by JustAnotherGhostName** [comment]
Can confirm it still occurs in 1.19.2. The corruption is way too frequent and also halts the server from cleanly shutting down. If you do /stop (dosnt matter if its console or not), the server will never fully shut down, and does not get to the saving state for the various dimensions. It ends up resulting in minor roll backs right before since the process has to be completely killed.

**Comment by togreglove** [comment]
We have resorted to running a /save-all by script just before shutdown on
any pack that uses FTBChunks, then issue an automated /kill 60 seconds
after it's supposed to have stopped.  I hate it, but it works till they
have time to address this.

**Comment by JustAnotherGhostName** [comment]
> utdow

That likely would work, however, combined with an issue in FTB Teams getting corrupted at nearly every other restart (https://github.com/FTBTeam/FTB-Mods-Issues/issues/449) im having to move away from allowing players to use either FTB Teams or FTB Chunks. There should not be a need for a server admin to manually intervene at nearly  EVERY restart of the server. I love the mods, but these corruptions are killing it.

**Comment by desht** [comment]
I'm going to a lot more info to proceed with this: "it still doesn't work" doesn't really help.  For a start, how do you know that it's actually even Teams or Chunks preventing a clean server shutdown when you run `/stop`.

In addition, Teams data getting corrupted does not necessarily indicate a bug with Teams.  If this were on every single shutdown and startup, I would expect a *lot* more people to be reporting problems right now.

**Comment by JustAnotherGhostName** [comment]
I also want to note that I do not know how to replicate it, at all. It just seems to occur on a world after a ~2 weeks weeks of new players joining and server activity. I only know that I can source it down to ftb chunks. I am currently in the process as well of swapping my users over to a different set of claiming and force loading methods, so at the very least, that will add another layer of confirmation as to whether or not ftb chunks causes the shutdown not to be clean.

**Comment by desht** [comment]
> I can confirm its teams data for both the corruption and loaded chunks by first going into ./server/world/data/chunks.dat with NBT explore and deleting the whole ftbchunks category of it, the unclaiming all chunks with the respective command.

The problem I have with all this right now is the conflation of Chunks and Teams here.  I'm disputing that a problem may exist, but your description of the problems seriously needs to be disentangled.

FTB Teams has exactly *nothing* to do with chunk loading, so I don't understand why you keep dragging it in to a discussion about chunk loading.

As for needing to clean out `chunks.dat` - I don't know.  That's just the forced chunk info that vanilla maintains, same as if you ran `/forceload` for the chunk.  FTB Chunks uses the Forge chunk manager to force/unforce chunks, and that just goes through the standard vanilla API for forcing/unforcing.  I cannot reproduce any issue where the server won't stop while there are forced chunks present in `chunks.dat` and I don't know what's happened to your world to cause that.

**Comment by JustAnotherGhostName** [comment]
I can write up a further description of just the FTB Chunks description from my end later today, however, my intention was just to further confirm other server admins are experiencing this issue and that its rather big. I dont think its a question as to whether or not this bug exists in FTB Chunks since this issue has been open for over a year, flagged as high priority, with multiple individuals explaining the issue exists.

As for FTB Teams, I originally only mentioned FTB Teams to highlight why the /save-all command and auto killing the process is not a valid solution in my case as it does not solve the FTB Chunks issue, and even if it did, FTB Teams still has the issue. My apologies for any confusion beyond that.

To strictly define the line between the two bugs, the FTB Chunks bug prevents the server from completely shutting down and just hangs there. As for FTB Teams, it prevents the server from completely booting up, stopping at loading world and not progressing further.

**Comment by desht** [comment]
> with multiple individuals explaining the issue exists.

Leave the anecdotes out, please.  If other people are experiencing this, they need to be confirming it, ideally in this thread.  If we actually knew what environment those who experience these problems are in , that might provide some useful clues as to what's causing them.  Because I know for a _fact_ that it's not everybody who gets these problems (and the fact that I simply can't reproduce either problem on a dedicated server doesn't make it easy or even possible to determine what's going on).

**Comment by JustAnotherGhostName** [comment]
>  If we actually knew what environment those who experience these problems are in , that might provide some useful clues as to what's causing them. 

Certainly, I understand that. it is a pain to diagnose the code side of things when you dont even know the steps to reproduce it in game. I cant say for certain when I can spend a few hours investigating this further than the time I took to track down the existence of this issue, however, I will try to replicate it in single player this week to see if there is any insights. I am guessing though that it might have some interaction with joining and leaving a FTB Team's party / team, but I may be wrong. It is also possible that the issue may become worse combined with the FTB Teams file rollbacks which become necessary at times, so I will try to incorporate that in my tests.

**Comment by SpottyTheTurtle** [comment]
I'm experiencing this on a 1.19.2 forge server, I can provide the mod list as necessary, it's a public server so I'm happy to help assist in trying to find the issue if needed.

**Comment by desht** [comment]
I've actually just committed another fix which I'm hoping will help; expect a new release in the next day or two.

**Comment by chrisfand** [comment]
Not sure if this was already accounted for in the fix, but if the server crashes while players were online with force loaded chunks, I noticed the chunks would be stuck forceloaded after restarting. So maybe a check is needed in that case to see if the player is online and, if not, it gets unloaded. This is on 1.16.5 with build 220.

**Comment by desht** [comment]
My fixes are focused on 1.18 and 1.19; 1.16 is in critical fixes only now.  However I will take a look at fixes for 1.16 when I get a chance.  The server crashing is likely to leave persistence files like `chunks.dat` in an inconsistent state, it's true.

**Comment by Jinakaks** [comment]
I can confirm that there is this issue exists in Medieval MC [FABRIC] MMC1 (version 18.5) (MC version 1.19.2)
as @chrisfand said the issue is related to the server crashing.

Steps to reproduce: (not yet verified, but highly likely)
1) force-load some chunks using FTB-chunks
2) crash the server (terminate the process)
3) restart the server

On our server, the chunks that were force loaded upon a server crash are stuck as pseudo-lazy chunks, no matter how close a player is.
Specifically, **the chunks will not tick blocks**. 
Interestingly, entities will tick fine in these chunks, and so will random ticks (crops grow fine). What doesn't work is furnaces, hoppers, water flowing, bamboo fully breaking, etc.

Additionally, these chunks can be force-loaded again, which will work fine for entities and random ticks, but no matter what I do I cannot get them to tick blocks again.

Also, if the server restarts normally while a new entity is put into one of these broken chunks, the entities are not saved to the chunk.

For example, the villagers in my villager breeder vanish at every restart. The chickens in my automatic chicken farm vanish on a restart. Both of these were in FTB force-loaded chunks upon a server crash, and both of these chunks don't tick blocks anymore. 

I assume what's happening is that the chunk is labeled as load level 32 or 33 (i.e. some level of lazy), so it assumes that it has already saved the entities contained within. Therefore the chunk doesn't save the entities, and so they simply vanish when the server closes. Some evidence for this theory is that my iron farm (in a broken chunk) has 2 villagers in it, and when I restock it with the third villager, that villager vanishes upon a server restart, but the original 2 remain there.

Evidently, the chunks.dat file doesn't exist in the server files so I'm not sure how to go about fixing the issue.

I wanted to make this post to help clarify the issue. I hope this helps!

**Comment by quartzar** [comment]
Encountering the same issue on Fabric 0.14.17, Minecraft version 1.19.2.

Deleting `/server/world/data/chunks.dat` does not fix the issue upon server restart.

This is what I see when I run `/ftbchunks admin view_loaded_chunks` and open up the map:

![image](https://user-images.githubusercontent.com/79379455/229165606-0cbe837c-df6b-4338-b5af-83e28f66b47f.png)

It persists after removal of FTB Chunks jar too. I'm seeing 4000-7000 chunks being loaded at any one time, irrespective of anybody being online.

**Comment by creaper33** [comment]
I'm experiencing this issue too. Deleting chunks.dat removed 1 bugged force-loaded base's force-loaded chunks, but the rest remained (odd). I would be willing to provide a server zip with most region files removed for size reasons.

**Comment by MichaelHillcox** [comment]
As this issue is very old and we've pushed a large amount of changes to the mod. Please check if your issue is still present in 1.21 versions of the mod. If it is, please open a new issue with updated information so it's top of our list 👍
